frontend/app/src/next/components/runs/run-dag/dag-run-visualizer.tsx

```diff
@@ -0,0 +1,239 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import ReactFlow, {
+  Position,
+  MarkerType,
+  Node,
+  Edge,
+  BezierEdge,
+} from 'reactflow';
+import 'reactflow/dist/style.css';
+import dagre from 'dagre';
+import { useTheme } from '@/next/components/theme-provider';
+import stepRunNode, { NodeData } from './step-run-node';
+import { V1TaskStatus } from '@/lib/api';
+import { RunDetailProvider, useRunDetail } from '@/next/hooks/use-run-detail';
+import { useNavigate } from 'react-router-dom';
+import { ROUTES } from '@/next/lib/routes';
+const connectionLineStyleDark = { stroke: '#fff' };
+const connectionLineStyleLight = { stroke: '#000' };
+
+const nodeTypes = {
+  stepNode: stepRunNode,
+};
+
+const edgeTypes = {
+  smoothstep: BezierEdge,
+};
+
+interface WorkflowRunVisualizerProps {
+  workflowRunId: string;
+  onTaskSelect?: (taskId: string) => void;
+}
+
+function WorkflowRunVisualizer({
+  workflowRunId,
+  onTaskSelect,
+}: WorkflowRunVisualizerProps) {
+  return (
+    <RunDetailProvider runId={workflowRunId}>
+      <WorkflowRunVisualizerContent
+        onTaskSelect={onTaskSelect}
+        workflowRunId={workflowRunId}
+      />
+    </RunDetailProvider>
+  );
+}
+
+function WorkflowRunVisualizerContent({
+  workflowRunId,
+  onTaskSelect,
+}: WorkflowRunVisualizerProps) {
+  const { theme } = useTheme();
+  const navigate = useNavigate();
+  const { data, isLoading, error } = useRunDetail();
+
+  const shape = useMemo(() => data?.shape || [], [data]);
+  const taskRuns = useMemo(() => data?.tasks || [], [data]);
+
+  const [containerWidth, setContainerWidth] = useState(0);
+  const containerRef = useRef(null);
+
+  useEffect(() => {
+    const resizeObserver = new ResizeObserver((entries) => {
+      for (const entry of entries) {
+        setContainerWidth(entry.contentRect.width);
+      }
+    });
+
+    if (containerRef.current) {
+      resizeObserver.observe(containerRef.current);
+    }
+
+    return () => {
+      if (containerRef.current) {
+        resizeObserver.unobserve(containerRef.current);
```

Comment:
The ref value 'containerRef.current' will likely have changed by the time this effect cleanup function runs. If this ref points to a node rendered by React, copy 'containerRef.current' to a variable inside the effect, and use that variable in the cleanup function